### PR TITLE
Improve coordinates display panel for spectrum viewer

### PR DESCRIPTION
**Pull request by pllim**
### Description

This pull request is to implement the following feature request:

* The display should show the pixel in x on the first line and spectral axis and flux on the second line with appropriate units.
* The cursor should lock to the spectrum in x only.

### TODO

- [x] Fix Cubeviz. It uses `PaddedSpectrumWCS` and not `SpectralCoordinates`.
- [x] Fix Mosviz.
- [x] Specviz2d? Need #1889 
- [x] Generalize the new code?
- [x] Make it work on collapsed Subset
- [x] Visualization like https://tradingeconomics.com/commodity/lumber minus the dashed lines. See https://github.com/pllim/jdaviz/pull/8
  - Is marker behavior acceptable? Is it too confusing to have both marker and the cursor?
  - Are different display formats acceptable when tool is toggled on and off?
- [ ] Acceptable performance (especially using MANGA cube in Cubeviz)

### Change log entry

- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`? If you want to avoid merge conflicts,
  list the proposed change log here for review and add to `CHANGES.rst` before merge. If no, maintainer
  should add a `no-changelog-entry-needed` label.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [x] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Did the CI pass? If not, are the failures related?
- [x] Is a milestone set? Set this to bugfix milestone if this is a bug fix and needs to be released ASAP; otherwise, set this to the next major release milestone.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)? [🐱](https://jira.stsci.edu/browse/JDAT-2947)

### CODE REVIEW

**Comment by pllim** [comment]
I might hold off on this until after #1908 is merged to avoid conflicts.

File: jdaviz/configs/default/plugins/gaussian_smooth/tests/test_gaussian_smooth.py
Comment (pllim) [inline jdaviz/configs/default/plugins/gaussian_smooth/tests/test_gaussian_smooth.py:15]: There is slight behavior change when you use the proper route to load the data into Cubeviz. I think this is more correct as a test because this is how user will actually load it.

File: jdaviz/configs/default/plugins/gaussian_smooth/tests/test_gaussian_smooth.py
Comment (pllim) [inline jdaviz/configs/default/plugins/gaussian_smooth/tests/test_gaussian_smooth.py:72]: I moved the warning check to the affected line. This way, we won't accidentally ignore warning that we're not supposed to ignore.

**Comment by pllim** [comment]
p.s. I actually think it is ok to merge this without the marker thingy, but I can also wait if everyone else wants the marker to be part of this PR.

File: jdaviz/configs/specviz/plugins/viewers.py
Comment (kecnry) [inline jdaviz/configs/specviz/plugins/viewers.py:97]: can you just use the arrays in the Lines object instead?  If that isn't accessible from the layer, we could instead loop through `self.figure.marks` (but would then need to filter out marks that are not corresponding to data entries, etc).

**Comment by kecnry** [comment]
The logic for selecting the closest spectrum doesn't behave as I'd expect.  See the following example (which uses https://github.com/pllim/jdaviz/pull/8 to display a marker at the currently exposed coordinates):

https://user-images.githubusercontent.com/877591/210006959-0b292af1-16d4-416a-8a95-ce6c8760ea58.mov

File: jdaviz/configs/specviz/plugins/viewers.py
Comment (pllim) [reply jdaviz/configs/specviz/plugins/viewers.py:97]: I can try using the marks if you tell me how to properly grab them. I get lost in all the spectrum viewer internals. 

Will using marks be compatible with unit conversion (when it finally works) and what is the risk of it breaking if we refactor the display stuff inside spectrum viewer?

**Comment by pllim** [comment]
> The logic for selecting the closest spectrum doesn't behave as I'd expect

I think I know what is happening. The distance math is pretty basic radius calculation using raw wave and flux values. So in the video, your mouse position results in distance in Y dominating the calculations (flux value is much higher than wave) and my math does not check if the X is outside a spectrum's wave bounds. So I think the fix shouldn't be hard. Good catch!

**Comment by pllim** [comment]
Re: https://github.com/spacetelescope/jdaviz/pull/1894#issuecomment-1367565592

Actually, I have trouble reproducing the problem you pointed out using dummy data. How are those spectra generated?

**Comment by pllim** [comment]
OK nvm... I just need to make `flux_2` much lower (60 instead of 1000).

File: jdaviz/configs/specviz/plugins/viewers.py
Comment (pllim) [reply jdaviz/configs/specviz/plugins/viewers.py:97]: Also, would using marks work for Cubeviz and Mosviz?

**Comment by pllim** [comment]
> display a marker at the currently exposed coordinates

I find it quite disorienting to see the jumpy marker and the mouse moving at different rates when I do mouseover. I wonder if @Jenneh has ideas.

File: jdaviz/configs/specviz/plugins/viewers.py
Comment (kecnry) [reply jdaviz/configs/specviz/plugins/viewers.py:97]: You should be able to pull the units from the figure as well, but will also need to account for `as_steps` similar to how is done for the cubeviz slider https://github.com/spacetelescope/jdaviz/blob/66adb92b48ea38d81a13251395cee94bc601e2ba/jdaviz/configs/cubeviz/helper.py#L103-L106

Either way, it's probably useful to cache in the plugin and update whenever the underlying data changes, but that might require a significant refactor of the plugin code.  Right now, cubeviz is very noticeably laggy, I suspect because of this line and the extra expense of re-collapsing and creating a `Spectrum1D` object at every mouse move event)


https://user-images.githubusercontent.com/877591/210076688-424f8439-c3d7-4d32-8fba-c3b83b6b33f6.mov

Comment (pllim) [reply jdaviz/configs/specviz/plugins/viewers.py:97]: In that case... I wonder if part of this can be refactored out so all the viz with spectrum viewer can reuse this same code. What do you think?

https://github.com/spacetelescope/jdaviz/blob/66adb92b48ea38d81a13251395cee94bc601e2ba/jdaviz/configs/cubeviz/helper.py#L89

**Comment by kecnry** [comment]
> I find it quite disorienting to see the jumpy marker and the mouse moving at different rates when I do mouseover

Does https://github.com/pllim/jdaviz/pull/8#issuecomment-1367958879 help at all (hiding the marker during zoom/subset actions)?  

My vote would be to include the marker for now since I think it adds a lot of important context to the mouseover display and avoids confusion/misinterpretation, and if it becomes annoying to us or users we can consider styling tweaks (color, marker, markersize, etc) and/or a toggle somewhere (perhaps directly in the space occupied by the mouseover coordinates since there isn't anything to display there when not hovering over a plot - see #1713 for a similar concept).

File: jdaviz/configs/specviz/plugins/viewers.py
Comment (kecnry) [reply jdaviz/configs/specviz/plugins/viewers.py:97]: I think the logic needs to be quite different here (can't just assume the first marks entry in specviz, for example).  But if there is any part that can be refactored to be shared, I'm all for that.  But with the complication of `as_steps`, I'm beginning to think just caching might be the cleanest way to go.

Comment (pllim) [reply jdaviz/configs/specviz/plugins/viewers.py:97]: The cleanest way to catch is probably deep upstream here though I have a feeling they won't want it because of extra memory usage...

https://github.com/glue-viz/glue/blob/2293019762136c8cda1ce67e9c89d5513013d7e7/glue/core/data_collection.py#L414

File: jdaviz/configs/specviz/plugins/viewers.py
Comment (pllim) [inline jdaviz/configs/specviz/plugins/viewers.py:485]: Pretty sure you want the underlying numpy array, not the Quantity right? I think the missing `.value` here might be a bug so I added it but please confirm. Thanks!

File: jdaviz/configs/specviz/plugins/viewers.py
Comment (pllim) [reply jdaviz/configs/specviz/plugins/viewers.py:94]: Yes... I forgot that Subset gets its own spectrum... sigh.

Comment (pllim) [reply jdaviz/configs/specviz/plugins/viewers.py:94]: It is gonna be more complicated because we allow people to edit Subset now. 🤪

Comment (kecnry) [reply jdaviz/configs/specviz/plugins/viewers.py:94]: right, that will need to clear/overwrite the cache... sigh

**Comment by Jenneh** [comment]
Screen Shot 2023-01-03 at 7 38 32 AM

This looks good to me. My only concern is that the "x=" is inside the white space that is left between the other labels and info. Can all that white space be made equal?

File: jdaviz/configs/specviz/plugins/viewers.py
Comment (kecnry) [reply jdaviz/configs/specviz/plugins/viewers.py]: Support for spatial subsets and logic for non-visible layers is now included.

https://user-images.githubusercontent.com/877591/210607972-5ccc437a-8d08-4584-8772-adb666e6b66f.mov

**Comment by kecnry** [comment]
Display has been updated to show cursor coordinates on the top row and move the index to the spectral axis row, and changing the marker to a vertical rectangle.  Any refactoring and ability to "lock" a layer will be deferred for follow-up efforts.


https://user-images.githubusercontent.com/877591/210865343-ae7f7e5e-1707-4e7a-8388-07f6ddf207e9.mov

**Review by rosteen — APPROVED** [review]
Looks good to me, the feature works and seems quite performant. I checked that it works for multiple spectra, including a collapsed spatial subset. The "closest spectrum" behavior seems intuitive for well separated spectra, although I can see how it might get annoying with overlapping spectra - but locking is a follow up. 

The only quibble I have is that I would prefer the marker (the blue rectangle) to be thinner, but that's not a blocker.

File: jdaviz/configs/imviz/plugins/coords_info/coords_info.py
Comment (bmorris3) [inline jdaviz/configs/imviz/plugins/coords_info/coords_info.py:86]: Isn't this statement always true? 😏

File: jdaviz/configs/specviz/plugins/viewers.py
Comment (bmorris3) [inline jdaviz/configs/specviz/plugins/viewers.py:197]: Will this logic ignore any masks associated with the spectrum, which could be accessible via `get_subset_object` rather than `get_object`?

File: jdaviz/configs/specviz/plugins/viewers.py
Comment (bmorris3) [inline jdaviz/configs/specviz/plugins/viewers.py:504]: Should we be explicit that `cls=Spectrum1D` here, or is that guaranteed elsewhere? When we support translation to/from `NDDataArray` from https://github.com/astropy/astropy/pull/14175, we should be prepared to be specific.

**Review by bmorris3 — APPROVED** [review]
This looks great! Leaving approval with a few comments.

File: jdaviz/configs/specviz/plugins/viewers.py
Comment (bmorris3) [inline jdaviz/configs/specviz/plugins/viewers.py:108]: Could we change `'spectrum-viewer'` to `self.jdaviz_app._default_spectrum_viewer_reference_name`?

File: jdaviz/configs/specviz/plugins/viewers.py
Comment (pllim) [reply jdaviz/configs/specviz/plugins/viewers.py:197]: I am not familiar with how mask is supposed to work in Jdaviz. What is the workflow I should try to answer this question?

File: jdaviz/configs/specviz/plugins/viewers.py
Comment (pllim) [reply jdaviz/configs/specviz/plugins/viewers.py:504]: Here, I merely cached the `get_object()` output. The actual call predates this PR, so addressing this is out of scope here.

File: jdaviz/configs/specviz/plugins/viewers.py
Comment (pllim) [reply jdaviz/configs/specviz/plugins/viewers.py:197]: Actually, now that I look at the diff again, these calls also predate the PR, so addressing this comment is out of scope here.

**Comment by pllim** [comment]
Screenshot 2023-01-19 111840

To me, would be nicer if the marker is semi-transparent or just a simple "x" but I should defer to @Jenneh on this in a follow-up effort after this PR is merged.

File: jdaviz/configs/specviz/plugins/viewers.py
Comment (pllim) [reply jdaviz/configs/specviz/plugins/viewers.py]: Done. Good catch!

**Review by pllim — COMMENTED** [review]
@kecnry did an awesome job polishing up this PR and adding much values to it while I was out. I am okay with the changes and I think I addressed remaining comments from @bmorris3 .

Given there are 2 approvals and work waiting on this, I am merging it. Thanks, all!

**Pull request merged**
